### ISSUE TITLE
Frozen mouse focus after trying to rebind buttons

### ISSUE THREAD

**Issue by xhalo32** [swaywm/sway#4201]
Sway version: 1.0
Package from arch community: 1.0-8
Configuration file should be irrelevant, but if you explicitly ask me I'll send my whole configuration.

Debug log not available, was not recording sway's output when the bug happened. Sway did not crash. The bug is so far not reproducible with short period of trying to trigger it.

A video showing what happened: https://photos.app.goo.gl/opLWRSZP6sfkR4XG8
Unexpected behavior explained in words: The mouse has focused in one area of an application and trying to change focus by changing the workspace or moving between tabbed windows did not change the mouse focus. Thus the focused part of the application got all the events from the mouse in other windows and on other workspaces.

Before I filmed the video, the mouse focus was set to one chromium window, and changing windows then clicking on my mouse took me back to the previous window which was freezing my mouse's focus.

This is not a duplicate of https://github.com/swaywm/sway/issues/2576 because moving the mouse, pressing buttons on the mouse or the keyboard did not change the focus.

I was trying to bind my mouse's side button (button 12) to function as the middle click. The following commands were run:

> Here I was trying to rebind the button with unfortunately no success.
> I got the name `MOUS9 DONGLE Mouse` from my old i3 config from before switching to sway.
> I got the commands to rebind a mouse button in sway from https://github.com/swaywm/sway/issues/3611
```
$ xinput --list
⎡ Virtual core pointer                      id=2  [master pointer  (3)]
⎜   ↳ Virtual core XTEST pointer                id=4  [slave  pointer  (2)]
⎜   ↳ xwayland-pointer:26                       id=6  [slave  pointer  (2)]
⎜   ↳ xwayland-relative-pointer:26              id=7  [slave  pointer  (2)]
⎣ Virtual core keyboard                     id=3  [master keyboard (2)]
    ↳ Virtual core XTEST keyboard               id=5  [slave  keyboard (3)]
    ↳ xwayland-keyboard:26                      id=8  [slave  keyboard (3)]
$ xinput set-button-map 'MOUS9 DONGLE Mouse' 1 2 3 4 5 6 7 8 9 10 11 2
unable to find device 'MOUS9 DONGLE Mouse'
$ xinput set-button-map 'xwayland-pointer:26' 1 2 3 4 5 6 7 8 9 10 11 2
$ xinput set-button-map 4 1 2 3 4 5 6 7 8 9 10 11 2
$ xinput set-button-map 6 1 2 3 4 5 6 7 8 9 10 11 2
$ xinput set-button-map 7 1 2 3 4 5 6 7 8 9 10 11 2
$ swaymsg 'bindsym --whole-window BTN_SIDE seat seat0 cursor press BTN_MIDDLE; seat seat0 cursor release BTN_MIDDLE'
$ swaymsg 'bindsym --whole-window 12 seat seat0 cursor press BTN_MIDDLE; seat seat0 cursor release BTN_MIDDLE'
Error: Unknown button 12
$ swaymsg 'bindsym --whole-window button12 seat seat0 cursor press BTN_MIDDLE; seat seat0 cursor release BTN_MIDDLE'
Error: Only buttons 1-9 are supported. For other mouse buttons, use the name of the event 
$ xev -event mouse
```

After the last command was run, the bug was triggered while using Chromium version `74.0.3729.157 Arch Linux` and trying to press the button 12 on my mouse. The exact details I have forgotten and have not been able to reproduce said bug. In the end I had to quit the sway session using `swaymsg 'exit'` after which sway worked like it had before.

I will try to reproduce this bug and will keep you informed with more exact details if I manage to trigger it again.

*Please ask for more details I am more than glad to help this project!*

EDIT (before posting the issue): I managed to reproduce the bug, but it happened about 10min after running the commands. I'll keep you updated!

**Comment by xhalo32** [comment]
Alright I managed to figure it out!

After running the command
```
bindsym --whole-window BTN_SIDE seat seat0 cursor press BTN_MIDDLE; seat seat0 cursor release BTN_MIDDLE
```
and pressing the back button (BTN_SIDE) on my mouse it sends a middle click to the active window and freezes the mouse focus. Once I press the middle mouse button the focus is unfrozen and I can use my applications again and even repeat the bug.

Here's a quick video demostrating what happened: https://drive.google.com/file/d/1J9e4WTGwffT9kGKWG45C2BtC7Dvzqor7/view?usp=sharing
> Before starting the screen recording, all I did was start sway with `sway -d 2> sway.log` and open a terminal (alacritty).

The command run in the video is the following
```
swaymsg 'bindsym --whole-window BTN_SIDE seat seat0 cursor press BTN_MIDDLE; seat seat0 cursor release BTN_MIDDLE'
```
It is supposed to completely rebind my back button to function as my middle button as per the fix in https://github.com/swaywm/sway/issues/3611. However once I press said button the mouse focus is frozen and I am not able to use my mouse to do pretty much anything else other than using that window or doing keyboard-only things in other windows.

It is possible that I'm using sway wrong, and my wish of rebinding my mouse buttons has not been fulfilled. If anyone knows of a working solution to rebind mouse buttons I'm happy to try it!

Oh, and the log file:
[swayclean.log](https://github.com/swaywm/sway/files/3244160/swayclean.log)

**Comment by RedSoxFan** [comment]
Can you please try the following:
```
swaymsg 'bindsym --whole-window BTN_SIDE "seat seat0 cursor press BTN_MIDDLE; seat seat0 cursor release BTN_MIDDLE"'
```

The quotes are needed when using a semicolon (`;`) or comma (`,`) as part of the command for a bindsym when adding it via swaymsg since they are command separators. So your original was running:
```
1. bindsym --whole-window BTN_SIDE seat seat0 cursor press BTN_MIDDLE
2. seat seat0 cursor release BTN_MIDDLE
```
The second part was executing immediately and causing the button state to become corrupt since `BTN_MIDDLE` was not pressed (related: https://github.com/swaywm/wlroots/issues/1077).

The quotes are not needed in the config file since the commands are read line by line and not split on the separators.

Hopefully that helps

**Comment by xhalo32** [comment]
The command does make my mouse's back button work as the middle mouse button, but every time I press it it completely freezes the focused window for ~2s.

I'll restart sway and send you the log file.

**Comment by xhalo32** [comment]
Alright here's the log: [sway2.log](https://github.com/swaywm/sway/files/3244727/sway2.log)

All I did after starting the debug sway session was
1. Open alacritty
2. In alacritty, run the command `swaymsg 'bindsym --whole-window BTN_SIDE "seat seat0 cursor press BTN_MIDDLE; seat seat0 cursor release BTN_MIDDLE"'`
3. Write dummy text and select it with my mouse
4. Press the back button on my mouse 4 times
5. Copy the log

Just to be clear, the 2 second freeze happens in every other application as well.

Hopefully you find something useful, I'm happy to help you!

**Comment by RedSoxFan** [comment]
Does the following also cause an issue?
```
bindsym --whole-window BTN_SIDE seat seat0 cursor press BTN_MIDDLE
bindsym --whole-window --release BTN_SIDE seat seat0 cursor release BTN_MIDDLE
```

**Comment by xuhdev** [comment]
@RedSoxFan I don't think it will work. See https://github.com/swaywm/sway/issues/5333#issuecomment-2667450431
